lib/libzfs/libzfs_util.c

```diff
@@ -862,7 +862,12 @@ libzfs_init(void)
 		return (NULL);
 	}
 
-	hdl->libzfs_sharetab = fopen("/etc/dfs/sharetab", "r");
+	if ((hdl->libzfs_sharetab = fopen("/etc/dfs/sharetab", "r")) == NULL) {
+		(void) close(hdl->libzfs_fd);
+		(void) fclose(hdl->libzfs_mnttab);
+		free(hdl);
+		return (NULL);
+	}
```

Comment:
It's probably better to remove this entirely from libzfs_init() / libzfs_close().  It's only used in   It's only ever used in `is_shared()` and it would be better to just handle the open/close there.